src/coreclr/jit/assertionprop.cpp

```diff
@@ -4367,6 +4367,60 @@ GenTree* Compiler::optAssertionProp_RelOp(ASSERT_VALARG_TP assertions, GenTree*
     return optAssertionPropLocal_RelOp(assertions, tree, stmt);
 }
 
+//--------------------------------------------------------------------------------
+// optVisitReachingAssertions: given a tree, call the specified callback function on all
+//    the assertions that reach it via PHI definitions if any.
+//
+// Arguments:
+//    GenTree    - The tree to visit all the reaching assertions for
+//    argVisitor - The callback function to call on the vn and its reaching assertions
+//
+// Return Value:
+//    AssertVisit::Aborted  - an argVisitor returned AssertVisit::Abort, we stop the walk and return
+//    AssertVisit::Continue - all argVisitor returned AssertVisit::Continue
+//
+template <typename TAssertVisitor>
+Compiler::AssertVisit Compiler::optVisitReachingAssertions(GenTree* tree, TAssertVisitor argVisitor)
+{
+    // It is better to rely only on VN and do what VNVisitReachingVNs does, but it is tricky
+    // to propagate BBs there. So, we do a simple walk here (no nested PHI nodes).
+    //
+    // We assume that the caller already checked aggregated assertions for the current block,
+    // so we're interested only in reaching assertions via PHI nodes.
+    //
+    if (!tree->OperIs(GT_LCL_VAR) || !tree->AsLclVar()->HasSsaName())
+    {
+        return AssertVisit::Abort;
+    }
+
+    GenTreeLclVarCommon* lcl    = tree->AsLclVarCommon();
+    LclSsaVarDsc*        ssaDef = lvaGetDesc(lcl->GetLclNum())->GetPerSsaData(lcl->GetSsaNum());
+    GenTreeLclVarCommon* node   = ssaDef->GetDefNode();
+
+    if ((node == nullptr) || !node->OperIs(GT_STORE_LCL_VAR) || !node->Data()->OperIs(GT_PHI))
+    {
+        return AssertVisit::Abort;
+    }
+
+    for (GenTreePhi::Use& use : node->Data()->AsPhi()->Uses())
+    {
+        GenTreePhiArg* phiArg     = use.GetNode()->AsPhiArg();
+        const ValueNum phiArgVN   = vnStore->VNConservativeNormalValue(phiArg->gtVNPair);
+        ASSERT_TP      assertions = optGetEdgeAssertions(ssaDef->GetBlock(), phiArg->gtPredBB);
+        if (BitVecOps::MayBeUninit(assertions))
+        {
+            return AssertVisit::Abort;
+        }
+
+        if (argVisitor(phiArgVN, assertions) == AssertVisit::Abort)
+        {
+            // The visitor wants to abort the walk.
+            return AssertVisit::Abort;
+        }
+    }
+    return AssertVisit::Continue;
```

Comment:
NOTE: we can walk nested PHIs (with some budget) or look for PHI defs via SSA longer than the just the first def, but that improved diffs only a tiny bit, so I didn't bother